src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringUtil.java

```diff
@@ -0,0 +1,74 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.script;
+
+import java.util.ArrayList;
+
+public class KNNScoringUtil {
+
+    public static float l2Squared(float[] queryVector, float[] inputVector) {
+        float squaredDistance = 0;
+        for (int i = 0; i < inputVector.length; i++) {
+            float diff = queryVector[i]-inputVector[i];
+            squaredDistance += diff * diff;
+        }
+        return squaredDistance;
+    }
+
+    public static float cosinesimilOptimized(float[] queryVector, float[] inputVector, double normQueryVector) {
+        double dotProduct = 0.0f;
+        double normInputVector = 0.0f;
+        if (normQueryVector == -1) {
+            throw new IllegalStateException("Normalized query vector cannot be negative");
+        }
+        for (int i = 0; i < queryVector.length; i++) {
+            dotProduct += queryVector[i] * inputVector[i];
+            normInputVector += inputVector[i] * inputVector[i];
+        }
+        // Divide by zero check
+        double normalizedProduct = normQueryVector * normInputVector;
+        if (normalizedProduct == 0 ) {
+            return Float.MIN_VALUE;
```

Comment:
this check can be saved for most vectors by try-catching the unlikely case from the division below